File: app/db/models/download_filter.py
"""
下载过滤配置数据模型
"""
from sqlalchemy import Column, Integer, String, Boolean, DateTime, DECIMAL, Text
from sqlalchemy.sql import func

from app.db.models.base import Base


class DownloadFilterSettings(Base):
    """下载过滤设置模型"""
    __tablename__ = 'download_filter_settings'

    id = Column(Integer, primary_key=True, autoincrement=True)
    # 文件大小过滤
    min_file_size_mb = Column(Integer, nullable=False, default=300, comment="最小文件大小(MB)")
    max_file_size_mb = Column(Integer, nullable=True, comment="最大文件大小(MB), null表示无限制")
    
    # 文件类型过滤
    allowed_extensions = Column(Text, nullable=True, comment="允许的文件扩展名，JSON格式")
    blocked_extensions = Column(Text, nullable=True, comment="禁止的文件扩展名，JSON格式")
    
    # 种子过滤
    min_seed_count = Column(Integer, nullable=True, default=1, comment="最小种子数")
    max_total_size_gb = Column(DECIMAL(10, 2), nullable=True, comment="种子总大小限制(GB)")
    
    # 智能过滤
    enable_smart_filter = Column(Boolean, nullable=False, default=True, comment="启用智能过滤")
    skip_sample_files = Column(Boolean, nullable=False, default=True, comment="跳过样本文件")
    skip_subtitle_only = Column(Boolean, nullable=False, default=True, comment="跳过仅字幕文件")
    media_files_only = Column(Boolean, nullable=False, default=False, comment="只保留媒体文件(视频+字幕)")
    include_subtitles = Column(Boolean, nullable=False, default=True, comment="包含字幕文件")
    
    # 系统字段
    is_active = Column(Boolean, nullable=False, default=True, comment="是否激活")
    created_at = Column(DateTime, nullable=False, server_default=func.now(), comment="创建时间")
    updated_at = Column(DateTime, nullable=False, server_default=func.now(), onupdate=func.now(), comment="更新时间")

    def __repr__(self):
        return f"<DownloadFilterSettings(id={self.id}, min_file_size_mb={self.min_file_size_mb})>"

File: frontend/src/apis/downloadFilter.ts
import { request } from '@/utils/requests'

// 类型定义
export interface FilterSettings {
  id?: number
  min_file_size_mb: number
  max_file_size_mb?: number | null
  allowed_extensions?: string[] | null
  blocked_extensions?: string[] | null
  min_seed_count?: number | null
  max_total_size_gb?: number | null
  enable_smart_filter: boolean
  skip_sample_files: boolean
  skip_subtitle_only: boolean
  media_files_only: boolean
  include_subtitles: boolean
  is_active?: boolean
  created_at?: string
  updated_at?: string
}

export interface TorrentFileInfo {
  name: string
  path: string
  size: number
  size_mb: number
  extension: string
  is_video: boolean
  is_sample: boolean
  is_subtitle: boolean
}

export interface MagnetFilterResult {
  should_download: boolean
  total_files: number
  filtered_files: number
  filtered_size_mb: number
  filter_reason: string
  files: TorrentFileInfo[]
}

export interface TorrentFilterResult {
  success: boolean
  message: string
  original_files: number
  filtered_files: number
  filtered_size_mb: number
  files: TorrentFileInfo[]
}

export interface FilterStatistics {
  total_filtered_torrents: number
  total_saved_space_gb: number
  most_common_filtered_types: string[]
}

export interface ApiResponse<T = any> {
  success: boolean
  message: string
  data?: T
}

// 过滤设置API
export const getFilterSettings = (): Promise<ApiResponse<FilterSettings>> => {
  return request.get('/download-filter/settings')
}

export const saveFilterSettings = (settings: Omit<FilterSettings, 'id' | 'is_active' | 'created_at' | 'updated_at'>): Promise<ApiResponse> => {
  return request.post('/download-filter/settings', settings)
}

export const getDefaultSettings = (): Promise<ApiResponse<FilterSettings>> => {
  return request.get('/download-filter/default-settings')
}

export const resetToDefaultSettings = (): Promise<ApiResponse> => {
  return request.post('/download-filter/reset-settings')
}

// 过滤测试API
export const testMagnetFilter = (magnetUrl: string): Promise<ApiResponse<MagnetFilterResult>> => {
  return request.post('/download-filter/test-magnet', { magnet_url: magnetUrl })
}

export const filterTorrent = (torrentHash: string): Promise<ApiResponse<TorrentFilterResult>> => {
  return request.post('/download-filter/filter-torrent', { torrent_hash: torrentHash })
}

// 统计信息API
export const getFilterStatistics = (): Promise<ApiResponse<FilterStatistics>> => {
  return request.get('/download-filter/statistics')
}

// 清理相关API
export const previewCleanup = (torrentHash: string): Promise<ApiResponse> => {
  return request.get(`/download-filter/preview-cleanup/${torrentHash}`)
}

export const cleanupTorrent = (torrentHash: string, dryRun: boolean = true): Promise<ApiResponse> => {
  return request.post(`/download-filter/cleanup-torrent/${torrentHash}`, null, {
    params: { dry_run: dryRun }
  })
}

export const cleanupAllTorrents = (category?: string, dryRun: boolean = true): Promise<ApiResponse> => {
  return request.post('/download-filter/cleanup-all', null, {
    params: { category, dry_run: dryRun }
  })
}